### PR TITLE
Add more BOLT optimizations

### PR DESCRIPTION
**Pull request by blyxyas**
This PR adds some more BOLT flags. I'm not sure why they were not added on the first place. I sadly cannot benchmark the changes on my machines (I'm having some problems). To not expand the build time too much, I only added the most influential on final times (inlining small functions and peepholes being the most influential).

As this is just a experiment, let's not assign anyone yet. Let's hope that this one PR improves runtimes. Improving performance is hard :/
r? ghost

### CODE REVIEW

**Comment by Kobzol** [comment]
FWIW, we tried the peephole flag several times, and it never showed any performance improvement. I don't think that we have tried the other flags.

**Comment by Kobzol** [comment]
I chose the flags that we use by asking the BOLT maintainers on Discord which flags would they suggest, and went with their suggestions :) But that was many months ago, so it's possible that there are new/better flags. One of the BOLT maintainers is suggesting new flags (https://github.com/rust-lang/rust/pull/119418), however this is currently waiting on the LLVM 18 release, which should happen in a few days.

**Comment by Kobzol** [comment]
Looks like this increases memory usage, but doesn't seem to improve perf.

File: src/tools/opt-dist/src/bolt.rs
Comment (AJIOB) [inline src/tools/opt-dist/src/bolt.rs:70]: Looks like you removed an incorrect comment.

File: src/tools/opt-dist/src/bolt.rs
Comment (AJIOB) [inline src/tools/opt-dist/src/bolt.rs:69]: Should we really need to remove this optimization?

**Review by AJIOB — CHANGES_REQUESTED** [review]

File: src/tools/opt-dist/src/bolt.rs
Comment (blyxyas) [reply src/tools/opt-dist/src/bolt.rs:69]: This change was not intended. I'll revert this line and change the flags based on feedback.

**Comment by blyxyas** [comment]
I'll change this slightly based on feedback. I think it's pretty weird that no optimizations at all are being displayed, being that inlining small functions tends to be pretty powerful. I'll do another benchmark with the fixed code (removing peepholes and restoring a flag I accidentally removed) and if that doesn't work, we'll just have to let this one die.

File: src/tools/opt-dist/src/bolt.rs
Comment (AJIOB) [inline src/tools/opt-dist/src/bolt.rs]: Is it useless?

Comment (blyxyas) [reply src/tools/opt-dist/src/bolt.rs]: Based on Kobzol comments:

> we tried the peephole flag several times, and it never showed any performance improvement

If it's already tried, I won't test it again on Rust's infra. (And sadly I'm having problems of benchmarking it in my own servers)

Comment (Kobzol) [reply src/tools/opt-dist/src/bolt.rs]: The Android maintainers (that use BOLT in their own Rust toolchain build) had some success with it, but we weren't able to replicate it.

**Comment by blyxyas** [comment]
Hmmm... I think that I'll close this, that time improvement isn't enough compared to the regressions. I'll try making that CI change we talked about in Zulip

**Pull request closed without merging**
